queue/abstract/driver/utube.lua

```diff
@@ -73,37 +131,120 @@ function method.normalize_task(self, task)
+        -- there can be no more than one task from each utube in a space.
+        pcall(self.space_ready.insert, self.space_ready, {id, utube})
+    end
+end
+
 -- put task in space
 function method.put(self, data, opts)
-    local max
-
-    -- Taking the maximum of the index is an implicit transactions, so it is
-    -- always done with 'read-confirmed' mvcc isolation level.
-    -- It can lead to errors when trying to make parallel 'put' calls with mvcc enabled.
-    -- It is hapenning because 'max' for several puts in parallel will be the same since
-    -- read confirmed isolation level makes visible all transactions that finished the commit.
-    -- To fix it we wrap it with box.begin/commit and set right isolation level.
-    -- Current fix does not resolve that bug in situations when we already are in transaction
-    -- since it will open nested transactions.
-    -- See https://github.com/tarantool/queue/issues/207
-    -- See https://www.tarantool.io/ru/doc/latest/concepts/atomic/txn_mode_mvcc/
-
-    if box.cfg.memtx_use_mvcc_engine and (not box.is_in_txn()) then
-        box.begin({txn_isolation = 'read-committed'})
-        max = self.space.index.task_id:max()
-        box.commit()
-    else
-        max = self.space.index.task_id:max()
+    local transaction_opts = {}
+    if box.cfg.memtx_use_mvcc_engine then
+        transaction_opts = {txn_isolation = 'read-committed'}
+    end
+
+    local commit_requirements = not box.is_in_txn()
+    if commit_requirements then
+        box.begin(transaction_opts)
     end
 
+    local max = self.space.index.task_id:max()
+
     local id = max and max[1] + 1 or 0
     local task = self.space:insert{id, state.READY, tostring(opts.utube), data}
+    if self.ready_space then
+        put_ready(self, task[1], task[3])
+    end
+
+    if commit_requirements then
+        box.commit()
+    end
+
     self.on_task_change(task, 'put')
     return task
 end
 
+local function take_ready(self)
+    local transaction_opts = {}
+    if box.cfg.memtx_use_mvcc_engine then
+        transaction_opts = {txn_isolation = 'read-committed'}
+    end
+
+    local commit_requirements = not box.is_in_txn()
+
+    while true do
+        if commit_requirements then
+            box.begin(transaction_opts)
+        end
+
+        local task_ready = self.space_ready.index.task_id:min()
+        if task_ready == nil then
+            if commit_requirements then
+                box.commit()
+            end
+            return
+        end
+
+        local id = task_ready[1]
+        local task = self.space:get(id)
+
+        if task[2] == state.READY then
+            local taken = self.space.index.utube:min{state.TAKEN, task[3]}
+
+            if taken == nil or taken[2] ~= state.TAKEN then
+                task = self.space:update(id, { { '=', 2, state.TAKEN } })
+
+                self.space_ready:delete(id)
+
+                if commit_requirements then
+                    box.commit()
```

Comment:
Three `box.commit`s smells a bit bad. I think you can rewrite this function a bit.